### ISSUE TITLE
Trying to run BPE on source texts 

### ISSUE THREAD

**Issue by mohammedayub44** [OpenNMT/OpenNMT-tf#242]
Hi,

I'm trying to run BPE method to on our custom dataset. How can we use the BPE vocabulary along with the pretrained embeddings (for now Fast text etc. ) to fine tune the model. Is this advisable or are there any considerations to be taken care of ? Since the pretrained embedding do not have sub word units I was not sure how helpful ti will be to try out this.

Thanks !
Mohammed Ayub

**Comment by guillaumekln** [comment]
Hi,

I don't think the pretrained embeddings will be helpful if their are defined at the word-level but the model is trained at the subword-level.

If you want to experiment with pretrained embeddings, it's certainly better to directly train them on your tokenized data (using fastText or word2vec).

**Comment by mohammedayub44** [comment]
@guillaumekln  That clarifies it. If I had to use just the BPE method without embedding, is there a CLI argument that I can pass to the `train_and_eval`. I was reading through the documentation and realized there is option to do it on-the-fly while training but was unclear what changes/steps are necessary to config file to get it running successfully. I would really appreciate if you can briefly lay it out (or an example) 

Thanks !
Mohammed Ayub

**Comment by guillaumekln** [comment]
To get started, I recommend applying BPE on your training files **outside** of OpenNMT-tf. This way you will clearly see on what the model will be trained.

**Comment by mohammedayub44** [comment]
@guillaumekln  Sure I used the jointly learn bpe scripts you have provided in the third_party folder. I have three files with me now: 

**2 Vocabulary files** (one for English and another for Spanish) - each line separated by its frequency count:
![image](https://user-images.githubusercontent.com/7122670/47435229-2b72fd80-d772-11e8-80ec-d7f2bfe3ba8b.png)

**1 output file** containing the sub-word units generated from training files provided :
![image](https://user-images.githubusercontent.com/7122670/47435366-73922000-d772-11e8-896b-3a11ee4c0068.png)

How do I pass these files no into the training model. My guess is, add paths to these files in the `.yml` file or use a command line argument ? was unclear on this on any other changes required.

Thanks !
Mohammed Ayub

**Comment by guillaumekln** [comment]
You can use `apply_bpe.py` to encode your training data.

However, this has nothing to do with OpenNMT-tf so far. Maybe you want to take a tour on the forum to learn or ask about data preparation: http://forum.opennmt.net

**Comment by mohammedayub44** [comment]
@guillaumekln 
My bad, I have now applied bpe to both my training files. I have `training_bpe32k.en` and `traning_bpe32k.es`. How can I pass these files to `train_and_eval` model ?

Thanks !
Mohammed Ayub

**Comment by mohammedayub44** [comment]
@guillaumekln  I successfully managed to create the BPE dataset, from some of the forum posts realized that tokenization -->learning the BPE is better. So I did the same. Now my model training seems to get stuck when I'm passing all the BPE datasets to `train_and_eval` 
Below are the details and the config.yml file I'm using:

1) Tokenized the train,eval,test data set for En-Es language pair using the below command: 
`onmt-tokenize-text --tokenizer OpenNMTTokenizer --tokenizer_config token_config.yml < train.en > token_train.en`
Using Tokenization Config:
`mode: aggressive`
`joiner_annotate: true`
`segment_numbers: true`
`segment_alphabet_change: true`

2) Learnt and Applied BPE to the tokenized data (generated from step 1) using:

Jointly learn BPE - 
` python learn_joint_bpe_and_vocab.py --input /home/ubuntu/mukund_onmt/token_train.en /home/ubuntu/mukund_onmt/token_train.es --output /home/ubuntu/mukund_onmt/mayub_try/bpe/output_bpe.txt --write-vocabulary /home/ubuntu/mukund_onmt/mayub_try/bpe/train_bpe.en /home/ubuntu/mukund_onmt/mayub_try/bpe/train_bpe.es -s 100000 --min-frequency 5`

Apply BPE to train, eval and test fro En-Es - 
` python apply_bpe.py -i /home/ubuntu/mukund_onmt/token_train.en -c /home/ubuntu/mukund_onmt/mayub_try/bpe/output_bpe.txt -o /home/ubuntu/mukund_onmt/mayub_try/bpe/bpe_train_100k.en --vocabulary /home/ubuntu/mukund_onmt/mayub_try/bpe/train_bpe.en`

3) Changed Source files to support On-the-fly Tokenization, steps as mentioned in the Online Usage Documentation section. (made changes and run the setup.py file to install)

4) Finally, Learning the model with applied BPE data: 
Below is the config file used:

`model_dir: Transformer_token_bpe_run10242018/`

`data:`
`  train_features_file: mayub_try/bpe/bpe_32000.en` //Applied BPE train file for EN
`  train_labels_file: mayub_try/bpe/bpe_32000.en`  //Applied BPE train file for ES
`  eval_features_file: mayub_try/bpe/bpe_dev_32000.en`   //Applied BPE eval file for EN
`  eval_labels_file: mayub_try/bpe/bpe_dev_32000.en`    //Applied BPE eval file for ES
`  source_words_vocabulary: mayub_try/bpe/output_bpe.txt`  // Output file from learn_joint_bpe_and_vocab.py
`  target_words_vocabulary: mayub_try/bpe/output_bpe.txt` // Output file from learn_joint_bpe_and_vocab.py
`  source_tokenizer_config: mayub_try/token_config.yml`  //Token config as in step 1 
`  target_tokenizer_config: mayub_try/token_config.yml`  //Token config as in step 1 

`train:`
`  save_checkpoints_steps: 5000`
`  keep_checkpoint_max: 3`
`  save_summary_steps: 5000`
`  train_steps: 100000`

`eval:`
`  external_evaluators: [BLEU,BLEU-detok]`

Ran this overnight, **model still stuck at first step** (as below): 

`trying to initialize from file mayub_try/bpe/output_bpe.txt is already initialized.`
`INFO:tensorflow:Done running local_init_op.`
`INFO:tensorflow:Saving checkpoints for 0 into mayub_try/NMTS_token_bpe_run10242018/model.ckpt.`
`2018-10-25 12:49:02.296343: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 2139599 of 12992009`
`2018-10-25 12:49:55.310710: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:135] Shuffle buffer filled.`
`INFO:tensorflow:loss = 309.33554, step = 1`



Appreciate any help !

Mohammed Ayub

**Comment by guillaumekln** [comment]
If you apply BPE directly on your input files, you **must not** use any settings documented in the [Tokenization](http://opennmt.net/OpenNMT-tf/tokenization.html) page. It will make things easier to get started.

Now you have your training files but your vocabulary is likely wrong. It should start with:

```bash
<blank>
<s>
</s>
```

Either add them manually or run `onmt-build-vocab` on your training files as documented in the quickstart.

For the stuck training, are you running low on system memory?

Please post additional such questions on the forum instead. Thanks.

**Comment by mohammedayub44** [comment]
@guillaumekln Thanks. For building the vocab I'm guessing I can just use the default argument like - `onmt-build-vocab --size 100000 --save_vocab src-vocab.txt bpe_32000.en` 

I can carry forward the conversation in the Forum. 

Mohammed Ayub
